File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.py
from collections import deque
def get_d_nums(num):
    result = set([])
    for i in range(1, int(num**(0.5))+1):
        if num % i == 0:
            result.add(i)
            result.add(num//i)
    return result
def get_m_num(num1, num2):
    d_num1 = get_d_nums(num1)
    d_num2 = get_d_nums(num2)
    return num1*num2/(max(d_num1&d_num2))
def solution(arr):
    arr = deque(list(set(arr)))
    # 두 수를 비교해서 최소 공배수 구하기
    # 해당 수와 다른 수를 다시 비교하기
    # 
    while len(arr) != 1:
        new_num = get_m_num(arr.popleft(), arr.popleft())
        arr.append(new_num)
    #print(arr)
    return arr[0]
        

File: 프로그래머스/2/12953. N개의 최소공배수/N개의 최소공배수.java
import java.util.*;

class Solution {
    public static void log(Object o){
        System.out.println(o);
    }
    public static int gcd(int a, int b){
        // (a > b) a % b이면 b가 최대 공약수 
        if (a<b){
            int temp = a;
            a = b;
            b = temp;
        }
        while(true){
            if(a % b == 0){
                return b;
            }
            int temp = b;
            b = a % b;
            a = temp;
        }
    }
    public int solution(int[] arr) {
        /*
        두 수의 최대 공배수 = A * B / 최대 공약수
        최대 공약수 = GCD(A, B)
        한 쌍의 최대 공약수를 구하고
        해당 수를 넣음
        */
        
        if(arr.length == 1){
            return arr[0];
        }
        int result = 1;
        int n = arr.length;
        for(int i=0; i<n; i++){
            result = (result * arr[i]) / gcd(result, arr[i]);
            //log(result);
        }
        
        return result;
    }
}